libp2p/connection.nim

```diff
@@ -0,0 +1,120 @@
+## Nim-LibP2P
+## Copyright (c) 2018 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+import chronos, options, chronicles
+import peerinfo,
+       multiaddress,
+       stream/lpstream,
+       peerinfo,
+       varint,
+       vbuffer
+
+const DefaultReadSize*: uint = 64 * 1024
+
+type
+  Connection* = ref object of LPStream
+    peerInfo*: PeerInfo
+    stream*: LPStream
+    observedAddrs*: Multiaddress
+
+proc newConnection*(stream: LPStream): Connection =
+  ## create a new Connection for the specified async reader/writer
+  new result
+  result.stream = stream
+
+method read*(s: Connection, n = -1): Future[seq[byte]] {.gcsafe.} =
+  result = s.stream.read(n)
```

Comment:
there's no need to use result, last value of expression is implicitly returned - using this feature allows the compiler to give better error messages:
```
method read(...): ... = 
  s.stream.read(n)
```
same for other kinds of expressions:
```
let x = 
  if a: 5
  else: 6
```

Reply:
I generally agree and in one liners like this, referring to `result` is probably unnecessary, but in larger code blocks, being explicit about your return values is not only prefered but should IMO be required. Being explicit about your intent goes a long way in terms of readability.

I'm curious tho what makes the compiler give better diagnostics without the explicit `result`? I'll experiment with it a bit more - thanks for the tip :)!

Reply:
```
proc f(x: int): int =
 if x> 4: 
  if x == 5:
   3
 else: 6
```

which if does else belong to? did all paths provide a value? if you touch `result`, the compiler won't complain about the missing one.

same with `new result` - if you're copypasting code around, it's easy to miss. the compiler can help you spot it with the object syntax that keeps the whole construction scoped - there are also options coming to warn of missing field inits if you use it - impossible to do if you build up result part by part (due to undecidability)

Reply:
for being explicit, there's also the classic `return x` - this leaves fewer possibilities to make mistakes - for example you won't assign twice to `result` when you fall off an indent level, though not as few as the implicit return.

Reply:
Ah, good points. `return` is a good option in this cases. Noted :)!

Reply:
this reminds me of one place where implicit return is not supported: stuff that goes through macro transformation - there are gaps in the async macros that prevent it from working (something about the compiler not giving enough type information and the macros not being powerful enough to figure it out, afair..)